### DISCUSSION TITLE
Conduit Address for setApprovalForAll

### DISCUSSION THREAD

**Discussion by adraffy** [ProjectOpenSea/seaport#580] [Q&A] [answered]
How do I derive the conduit address for an owner so I can provide pre-approval for selling on OpenSea during my minting process?

On Rinkeby,  when I (`0x51050ec063d393217B436747617aD1C2285Aeeee`) sell an ERC721 (`0xa7b92650cb392fbefcb8fc02e32babde854742f1`) on `testnets.opensea.io` for the first time, it wants approval for `0x1E0049783F008A0085193E00003D00cd54003c71` which appears to be a conduit.

From the ConduitController, I can reverse that conduit to a key `0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000`

How do I derive this key?  I assume it is tied to an owner + the marketplace (OpenSea in this example)

I see that key is provided in JSON from a GraphQL request for `CreateListingActionModalQuery` which also include a Zone (`0x00000000E88FE2628EbC5DA81d2b3CeaD633E89e`)

Is there a reason none of this documented?

Any help would greatly appreciated.  Thanks.

**Comment by adraffy** [comment]
According to [this tweet](https://twitter.com/opensea_support/status/1540343956738670592) the mainnet conduit address is also `0x1E0049783F008A0085193E00003D00cd54003c71`.

**Comment by Slokh** [comment, marked as answer]
There is no per-user proxies anymore, all OpenSea users will use `0x1E0049783F008A0085193E00003D00cd54003c71` as the Conduit.

**Reply by syffs** [reply]
Could you please tell whether you advise future NFT contract to include something as below for gas free listing:
```solidity
function isApprovedForAll(address owner, address operator)
	public
	view
	virtual
	override
	returns (bool)
{
	if (operator == 0x1E0049783F008A0085193E00003D00cd54003c71 ) {
		return true;
	}

	return super.isApprovedForAll(owner, operator);
}
```

**Reply by adraffy** [reply]
In general, for any address that's you want to pre-approved, you should probably use a negated approval mapping.

Something like:
```
isPreapproved(address) {
    return address == 0x1E0049783F008A0085193E00003D00cd54003c71;
}

setApprovalForAll(operator, approved) {
    if (isPreapproved(addr)) {
        disapprovals[sender][operator] = !approved;
    } else {
        approvals[sender][operator] = approved;
    }
}

isApprovedForAll(owner, operator) {
    if (isPreapproved(addr)) {
        return !disapprovals[owner][operator];
    } else {
        return approvals[owner][operator];
    }
}
```

**Reply by syffs** [reply]
I agree, did that in a hurry. 
My point is, if we're supposed to pre-approve it, then
- [this](https://github.com/ProjectOpenSea/opensea-creatures/blob/master/contracts/ERC721Tradable.sol) should be updated
- to be backward compatible with all the contracts that included pre-approval through proxy registry, proxy registry should be updated to return `0x1E0049783F008A0085193E00003D00cd54003c71` as `proxies` for everyone. You cannot claim to be improving gas with seaport if you're breaking gas free listing for all contracts that used to support it.

**Reply by adraffy** [reply]
Agreed.
